### PR TITLE
refactor(spanner): reorganize backup_integration_test

### PR DESCRIPTION
**Pull request by devbww**
Refactor the `BackupTest` class, mostly in preparation for adding
new PITR and CMEK backup/restore tests.

Cleanup some handling of timestamps by using `absl::Time` for all
time computation, and `spanner::Timestamp` for all expectations.
Remove the assertion that "the server only preserves micros" as the
loss of precision is really due to funneling through `time_point`.

Also only run `CleanupStaleInstances()` once when the test starts,
rather than before every test case.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5868)

### CODE REVIEW

File: google/cloud/spanner/integration_tests/backup_integration_test.cc [hunk 1]
```diff
@@ -36,116 +37,115 @@ namespace spanner {
...
+    // Make sure we're using a correct regex.
+    EXPECT_EQ(2, instance_name_regex.mark_count());
+    auto generator = google::cloud::internal::MakeDefaultPRNG();
+    Instance in(ProjectId(), spanner_testing::RandomInstanceName(generator));
+    auto fq_instance_name = in.FullName();
+    std::smatch m;
+    EXPECT_TRUE(std::regex_match(fq_instance_name, m, instance_name_regex));
+    EXPECT_EQ(3, m.size());
+
+    if (RunSlowBackupTests()) {
```
Comment (coryan) [inline google/cloud/spanner/integration_tests/backup_integration_test.cc:77]: You can probably perform this cleanup regardless of whether slow tests are enabled or not, up to you.

File: google/cloud/spanner/integration_tests/backup_integration_test.cc [hunk 2]
```diff
@@ -36,116 +37,115 @@ namespace spanner {
...
+      spanner_testing::TestBackoffPolicy(),
+      spanner_testing::TestPollingPolicy()));
 
-  Database db(project_id_, instance_id, database_id);
-
-  auto database_future = database_admin_client_.CreateDatabase(db);
-  auto database = database_future.get();
+  Instance in(ProjectId(), spanner_testing::RandomInstanceName(generator_));
+  auto instance_config = spanner_testing::PickInstanceConfig(
+      in.project_id(), std::regex(".*us-west.*"), generator_);
```
Comment (coryan) [inline google/cloud/spanner/integration_tests/backup_integration_test.cc:114]: Maybe a comment to document why `us-west` is special?  I have some vague recollection that we have more room to create instances there or something.

File: google/cloud/spanner/integration_tests/backup_integration_test.cc [hunk 3]
```diff
@@ -36,116 +37,115 @@ namespace spanner {
...
+  ASSERT_FALSE(instance_config.empty()) << "could not get an instance config";
+  EXPECT_STATUS_OK(
+      instance_admin_client
+          .CreateInstance(CreateInstanceRequestBuilder(in, instance_config)
```
Comment (coryan) [inline google/cloud/spanner/integration_tests/backup_integration_test.cc:118]: Do you know why this test needs a new instance?  I mean, it is going to create databases and backups with random names, an existing instance should be able to do that?  Or is there some limitation on the number of backups per instance we need to workaround?

**Review by coryan — APPROVED** [review]
LGTM. There is some stuff in this test that is failure obscure, maybe I just don't remember why we made the decisions we made anymore.  Consider some comments (left some ideas below).

File: google/cloud/spanner/integration_tests/backup_integration_test.cc (hunk 1 above)
Comment (devbww) [reply google/cloud/spanner/integration_tests/backup_integration_test.cc:77]: I'll leave it as is for now, but will reassess when I return here soon for the PITR/CMEK tests.

File: google/cloud/spanner/integration_tests/backup_integration_test.cc (hunk 2 above)
Comment (devbww) [reply google/cloud/spanner/integration_tests/backup_integration_test.cc:114]: https://github.com/googleapis/google-cloud-cpp-spanner/pull/1403 appears to be the origin, but doesn't list a reason.

File: google/cloud/spanner/integration_tests/backup_integration_test.cc (hunk 3 above)
Comment (devbww) [reply google/cloud/spanner/integration_tests/backup_integration_test.cc:118]: No, I don't know.  My thought is there was no better way to ensure a "us-west" instance (for whatever reason that was).

I am working on a filtering capability for `PickRandomInstance()`, which will be required for the CMEK merge (at least for the tests in their current state).  See #5588.  At that time I can also simplify this (if we believe "us-west" is still a feature).

**Pull request merged**